File: tech-repo-scaffold/django-backbone/scripts/evidence/collect.py
#!/usr/bin/env python3
"""
ABR Evidence Collector
======================
Collects CI/test artifacts into an artifacts.json manifest with SHA-256
hashes.  Mirrors the NzilaOS evidence pipeline (collect → seal → verify).  # noqa: RUF001
# cSpell:ignore Nzila sarif cyclonedx

Usage:
    python scripts/evidence/collect.py [--output <dir>]

Outputs:
    <output>/artifacts.json  — manifest of hashed artifact files

Environment:
    ARTIFACT_DIR   Override base artifact search directory (default: .)
"""
import argparse
import hashlib
import json
import logging
import os
import sys
from pathlib import Path
from typing import Any

logging.basicConfig(level=logging.INFO, format="%(levelname)s  %(message)s")
log = logging.getLogger(__name__)

# Candidate artifact paths (relative to project root or common CI output dirs).
# Missing files are skipped (not an error); their absence IS noted in the manifest.
ARTIFACT_CANDIDATES: list[tuple[str, str]] = [
    # name                         relative path
    ("pip-audit-report", "audit-report.json"),
    ("pip-audit-report-ci", "ci-outputs/audit-report.json"),
    ("trivy-sarif", "trivy-results.sarif"),  # cSpell:ignore sarif
    ("trivy-sarif-ci", "ci-outputs/trivy-results.sarif"),
    ("sbom-cyclonedx", "sbom.json"),  # cSpell:ignore cyclonedx
    ("sbom-cyclonedx-ci", "ci-outputs/sbom.json"),
    ("django-migration-check", "ci-outputs/migration-check.txt"),
    ("pytest-results", "ci-outputs/pytest-results.xml"),
    ("pytest-coverage", "ci-outputs/coverage.xml"),
    ("bandit-report", "ci-outputs/bandit-report.json"),
    ("safety-report", "ci-outputs/safety-report.json"),
]


def sha256_file(path: Path) -> str:
    h = hashlib.sha256()
    with path.open("rb") as fh:
        for chunk in iter(lambda: fh.read(65536), b""):
            h.update(chunk)
    return h.hexdigest()


def collect(base_dir: Path, output_dir: Path) -> dict[str, Any]:
    found: list[dict[str, Any]] = []
    missing: list[str] = []

    for name, rel_path in ARTIFACT_CANDIDATES:
        candidate = base_dir / rel_path
        if candidate.exists():
            digest = sha256_file(candidate)
            entry: dict[str, Any] = {
                "name": name,
                "path": rel_path,
                "sha256": digest,
                "size_bytes": candidate.stat().st_size,
            }
            found.append(entry)
            log.info("  ✓  %-35s  %s…", name, digest[:16])
        else:
            missing.append(name)
            log.debug("  –  %-35s  (not found, skipping)", name)

    manifest: dict[str, Any] = {
        "schema_version": "1",
        "generator": "abr-evidence-collect",
        "artifacts": found,
        "missing_artifacts": missing,
    }

    output_dir.mkdir(parents=True, exist_ok=True)
    out_path = output_dir / "artifacts.json"
    out_path.write_text(json.dumps(manifest, indent=2), encoding="utf-8")
    log.info("Wrote %d artifacts to %s", len(found), out_path)
    return manifest


def main() -> None:
    parser = argparse.ArgumentParser(description="Collect ABR CI evidence artifacts")
    parser.add_argument("--output", default="evidence-output", help="Output directory")
    parser.add_argument(
        "--base", default=os.environ.get("ARTIFACT_DIR", "."), help="Base search dir"
    )
    args = parser.parse_args()

    base_dir = Path(args.base).resolve()
    output_dir = Path(args.output).resolve()

    log.info("ABR Evidence Collector")
    log.info("  base_dir   : %s", base_dir)
    log.info("  output_dir : %s", output_dir)

    result = collect(base_dir, output_dir)
    n_found = len(result["artifacts"])
    n_missing = len(result["missing_artifacts"])
    log.info("Collection complete — %d found, %d skipped", n_found, n_missing)

    if n_found == 0:
        log.error("No artifacts collected — evidence pack will be empty.")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: apps/union-eyes/scripts/evidence/collect.mjs
#!/usr/bin/env tsx
/**
 * Union Eyes — Evidence Artifact Collector
 *
 * Scans the CI workspace for security scan outputs, test reports, and SBOM
 * files, then writes a manifest (artifacts.json) for the seal step.
 *
 * Usage:
 *   pnpm tsx scripts/evidence/collect.mjs [--output <dir>]
 *
 * Produces:
 *   <output>/artifacts.json  — ordered list of artifact descriptors
 */
// @ts-check
import { createHash } from 'node:crypto'
import { existsSync, readFileSync, statSync, mkdirSync, writeFileSync } from 'node:fs'
import { join, resolve } from 'node:path'
import { parseArgs } from 'node:util'

// ── CLI args ─────────────────────────────────────────────────────────────────
const { values } = parseArgs({
  args: process.argv.slice(2),
  options: {
    output: { type: 'string', default: 'evidence-output' },
    'base-path': { type: 'string', default: process.cwd() },
  },
  strict: false,
})

const OUTPUT_DIR = resolve(typeof values.output === 'string' ? values.output : 'evidence-output')
const BASE_PATH = resolve(typeof values['base-path'] === 'string' ? values['base-path'] : process.cwd())

// ── Known artifact locations (produced by CI steps) ──────────────────────────
const ARTIFACT_CANDIDATES = [
  { path: 'audit-report.json', category: 'dependency-audit', name: 'pnpm-audit' },
  { path: 'audit-report-high.json', category: 'dependency-audit', name: 'pnpm-audit-high' },
  // cSpell:ignore cyclonedx sarif
  { path: 'sbom.json', category: 'sbom', name: 'cyclonedx-sbom' },
  { path: 'trivy-fs.sarif', category: 'trivy-scan', name: 'trivy-filesystem' },
  { path: 'trivy-results.sarif', category: 'trivy-scan', name: 'trivy-container' },
  // UE-specific test reports
  { path: 'test-results.json', category: 'unit-tests', name: 'ue-unit-tests' },
  { path: 'coverage/lcov.info', category: 'test-coverage', name: 'ue-coverage' },
]

// ── Hash a file ────────────────────────────────────────────────────────────────
function sha256File(/** @type {string} */ filePath) {
  const content = readFileSync(filePath)
  return createHash('sha256').update(content).digest('hex')
}

// ── Collect ────────────────────────────────────────────────────────────────────
function collect() {
  mkdirSync(OUTPUT_DIR, { recursive: true })

  const collected = []
  const missing = []

  for (const candidate of ARTIFACT_CANDIDATES) {
    const fullPath = join(BASE_PATH, candidate.path)
    if (!existsSync(fullPath)) {
      missing.push(candidate.name)
      continue
    }

    const stat = statSync(fullPath)
    const sha256 = sha256File(fullPath)

    collected.push({
      name: candidate.name,
      category: candidate.category,
      sha256,
      path: candidate.path,
      sizeBytes: stat.size,
      collectedAt: new Date().toISOString(),
    })

    console.log(`  ✅ Collected: ${candidate.name} (${sha256.slice(0, 16)}...)`)
  }

  if (missing.length > 0) {
    console.log(`  ⚠️  Missing (optional): ${missing.join(', ')}`)
  }

  const manifest = {
    version: '1',
    collectedAt: new Date().toISOString(),
    source: 'union-eyes',
    runId: process.env.GITHUB_RUN_ID ?? 'local',
    commitSha: process.env.GITHUB_SHA ?? 'unknown',
    artifacts: collected,
  }

  const outPath = join(OUTPUT_DIR, 'artifacts.json')
  writeFileSync(outPath, JSON.stringify(manifest, null, 2), 'utf-8')
  console.log(`\n📦 Artifact manifest written: ${outPath}`)
  console.log(`   Collected ${collected.length} artifact(s).`)

  return manifest
}

collect()
